File: packages/opencode/src/voice/record.ts
/**
 * Voice recording — capture mic audio via the bundled Python recorder.
 *
 * Uses sounddevice (already present in most scientific Python envs)
 * because Node/Bun has no cross-platform mic API without native deps.
 *
 * Resolution order for the Python executable:
 *   MOMO_VOICE_PYTHON → MOMO_SIM_PYTHON → "python"
 *
 * @module voice/record
 */

import { spawn } from "child_process"
import * as path from "path"
import { fileURLToPath } from "url"

export interface RecordResult {
  readonly path: string
  readonly seconds: number
  readonly peak: number
}

/** Default path of the bundled recorder script. */
export function defaultRecorderPath(): string {
  const here = path.dirname(fileURLToPath(import.meta.url))
  return path.resolve(here, "..", "..", "python", "voice", "record.py")
}

function resolvePython(): string {
  return (
    process.env.MOMO_VOICE_PYTHON ||
    process.env.MOMO_SIM_PYTHON ||
    "python"
  )
}

/**
 * Record `seconds` of microphone audio to `outPath` (WAV, 16kHz mono).
 * Throws with an actionable message on failure.
 */
export function recordAudio(
  outPath: string,
  seconds: number,
): Promise<RecordResult> {
  const python = resolvePython()
  const script = process.env.MOMO_VOICE_RECORDER || defaultRecorderPath()

  return new Promise((resolve, reject) => {
    const child = spawn(python, ["-u", script, outPath, String(seconds)], {
      env: process.env,
      stdio: ["ignore", "pipe", "pipe"],
      windowsHide: true,
    })

    let stdout = ""
    let stderr = ""
    const timer = setTimeout(
      () => {
        child.kill()
        reject(new Error(`recording timed out after ${seconds + 15}s`))
      },
      (seconds + 15) * 1000,
    )

    child.stdout?.on("data", (d: Buffer) => (stdout += d.toString()))
    child.stderr?.on("data", (d: Buffer) => (stderr += d.toString()))
    child.on("error", (err) => {
      clearTimeout(timer)
      reject(
        new Error(
          `failed to start python recorder: ${err.message} (set MOMO_VOICE_PYTHON)`,
        ),
      )
    })
    child.on("close", (code) => {
      clearTimeout(timer)
      if (code !== 0) {
        reject(
          new Error(
            `recorder exited with code ${code}: ${stderr.trim() || stdout.trim()}`,
          ),
        )
        return
      }
      try {
        const info = JSON.parse(stdout.trim().split("\n").pop() || "{}")
        resolve({
          path: info.path || outPath,
          seconds: info.seconds ?? seconds,
          peak: info.peak ?? 0,
        })
      } catch {
        reject(new Error(`recorder returned unparseable output: ${stdout.slice(0, 200)}`))
      }
    })
  })
}

File: packages/opencode/python/voice/record.py
#!/usr/bin/env python3
"""
voice record — capture microphone audio to a WAV file.

Usage:
    python record.py <output.wav> [seconds]

Records mono 16 kHz int16 WAV (Whisper-optimal, compact) using
sounddevice + scipy. Prints a single JSON line to stdout:

    {"ok": true, "path": "...", "seconds": 5, "frames": 80000}

Errors go to stderr with a non-zero exit code.
"""

import json
import sys


def main():
    if len(sys.argv) < 2:
        print("usage: record.py <output.wav> [seconds]", file=sys.stderr)
        sys.exit(2)

    out_path = sys.argv[1]
    seconds = float(sys.argv[2]) if len(sys.argv) > 2 else 5.0
    sample_rate = 16000

    try:
        import sounddevice as sd
    except ImportError:
        print(
            "sounddevice not installed — run: pip install sounddevice scipy",
            file=sys.stderr,
        )
        sys.exit(3)

    try:
        import numpy as np
        from scipy.io import wavfile
    except ImportError:
        print(
            "numpy/scipy not installed — run: pip install sounddevice scipy",
            file=sys.stderr,
        )
        sys.exit(3)

    frames = int(sample_rate * seconds)
    audio = sd.rec(frames, samplerate=sample_rate, channels=1, dtype="int16")
    sd.wait()

    # Skip files that are essentially silence (device failure heuristic)
    peak = int(np.abs(audio).max()) if audio.size else 0

    wavfile.write(out_path, sample_rate, audio)
    print(
        json.dumps(
            {"ok": True, "path": out_path, "seconds": seconds, "frames": frames, "peak": peak}
        )
    )


if __name__ == "__main__":
    main()
